timelock/src/lib.rs

```diff
@@ -0,0 +1,198 @@
+}
+
+#[derive(Clone)]
+#[contracttype]
+pub enum DataKey {
+    Balance,
+    Nonce(Identifier),
+}
+
+#[derive(Clone)]
+#[contracttype]
+pub enum TimeBoundKind {
+    Before,
+    After,
+}
+
+#[derive(Clone)]
+#[contracttype]
+pub struct TimeBound {
+    pub kind: TimeBoundKind,
+    pub timestamp: u64,
+}
+
+#[derive(Clone)]
+#[contracttype]
+pub struct ClaimableBalance {
+    pub token: BytesN<32>,
+    pub amount: BigInt,
+    pub claimants: Vec<Identifier>,
+    pub time_bound: TimeBound,
+}
+
+pub struct ClaimableBalanceContract;
+
+// The 'timelock' part: check that provided timestamp is before/after
+// the current ledger timestamp.
+fn check_time_bound(env: &Env, time_bound: &TimeBound) -> bool {
+    let ledger_timestamp = env.ledger().timestamp();
+
+    match time_bound.kind {
+        TimeBoundKind::Before => ledger_timestamp <= time_bound.timestamp,
+        TimeBoundKind::After => ledger_timestamp >= time_bound.timestamp,
+    }
+}
+
+// Contract usage pattern (pseudocode):
+// 1. Depositor calls `token.approve(depositor_auth, claimable_balance_contract, 100)`
+//    to allow contract to withdraw the needed amount of token.
+// 2. Depositor calls `deposit(depositor_auth, token_id, 100, claimants, time bound)`. Contract
+//    withdraws the provided token amount and stores it until one of the claimants
+//    claims it.
+// 3. Claimant calls `claim(claimant_auth)` and if time/auth conditons are passed
+//    receives the balance.
+#[contractimpl]
+impl ClaimableBalanceContract {
+    pub fn deposit(
+        env: Env,
+        from: Signature,
+        token: BytesN<32>,
+        amount: BigInt,
+        claimants: Vec<Identifier>,
+        time_bound: TimeBound,
+    ) {
+        if claimants.len() > 10 {
+            panic!("too many claimants");
+        }
+        if is_initialized(&env) {
+            panic!("contract has already been initialized");
+        }
+        let from_id = from.get_identifier(&env);
+        // Authenticate depositor with nonce of zero, so that this may
+        // be successfully called just once.
+        check_auth(
+            &env,
+            &NonceForSignature(from),
+            BigInt::zero(&env),
+            Symbol::from_str("deposit"),
+            (&from_id, &token, &amount, &claimants, &time_bound).into_val(&env),
+        );
+        // Transfer token to this contract address.
+        transfer_from(&env, &token, &from_id, &get_contract_id(&env), &amount);
+        // Store all the necessary balance to allow one of the claimants to claim it.
+        env.contract_data().set(
+            DataKey::Balance,
+            ClaimableBalance {
+                token,
+                amount,
+                time_bound,
+                claimants,
+            },
+        );
+    }
+
+    pub fn claim(env: Env, claimant: Signature) {
+        if !is_initialized(&env) {
+            panic!("contract hasn't been initialized");
+        }
```

Comment:
This check looks unnecessary. The following line will fail if there is no balance deposited.

Reply:
Removed the check.